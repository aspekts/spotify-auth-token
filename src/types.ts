export type OptionalConfig = {
  port: number;
  scopes: string;
};

export type RequiredConfig = {
  client_id: string;
  client_secret: string;
};

export type CLIInputConfig = RequiredConfig & Partial<OptionalConfig>;
export type Config = RequiredConfig & OptionalConfig;

export type SpotifyTokenResponse = {
  access_token: string;
  token_type: 'Bearer';
  expires_in: number;
  refresh_token: string;
  scope: string;
  date_obtained: string;
};

export type SpotifyAuthResponse = {
  code: string;
  state: string;
};
