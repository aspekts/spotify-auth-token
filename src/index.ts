import { authorize } from './authorize';

export type { UserConfig } from './types';
export default authorize;

module.exports = authorize;
