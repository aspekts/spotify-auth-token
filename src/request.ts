import http from 'http';
import https, { RequestOptions } from 'https';

export const getLocalhostUrl = async (port: number): Promise<string> => {
  return new Promise((resolve, reject) => {
    const server = http
      .createServer((req, res) => {
        res.end('You can now close this window');
        res.once('finish', () => {
          server.close();
          if (req.url) {
            resolve(req.url.slice(1));
          }
          reject("Couldn't get code or state");
        });
      })
      .listen(port);
  });
};

export async function request<T>(
  opts: RequestOptions,
  body: string
): Promise<T> {
  return new Promise((resolve, reject) => {
    const data: Uint8Array[] = [];
    const req = https.request(opts, (res) => {
      res.on('data', (chunk) => data.push(chunk));
      res.on('end', () => {
        const resString = Buffer.concat(data).toString();
        resolve(JSON.parse(resString));
      });
    });

    req.on('error', (e) => {
      reject(e);
    });

    req.write(body);
    req.end();
  });
}
