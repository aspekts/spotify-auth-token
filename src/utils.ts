import { existsSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';

export function write(path: string, fileName: string, data: unknown): void {
  if (!existsSync(path)) {
    mkdirSync(path, { recursive: true });
  }
  path = join(path, fileName + '.json');
  writeFileSync(path, JSON.stringify(data, null, 2));
}

export function id(): string {
  return Math.random().toString(36).slice(2);
}

export function goodBye(message: string): never {
  console.error('\x1b[31m', `Error: ${message}. Goodbye`, '\x1b[0m');
  process.exit(1);
}
