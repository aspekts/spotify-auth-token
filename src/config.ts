import { UserConfig } from './index';
import { AppConfig } from './types';
import { goodBye, hasOwnProperty } from './utils';

const defaultConfig: AppConfig = {
  clientId: '',
  clientSecret: '',
  port: 3000,
  outDir: process.cwd(),
  outFileName: 'spotify-token',
  scopes: 'user-read-email',
};

export const createConfig = (args: UserConfig | string[]): AppConfig => {
  const config: AppConfig = { ...defaultConfig };

  // CLI use
  if (Array.isArray(args)) {
    args.forEach((val, idx, orig) => {
      if (val.startsWith('--')) {
        const key = val.slice(2);
        if (hasOwnProperty(config, key)) {
          config[key] = orig[idx + 1];
        }
      }
    });

    // Programmatic use
  } else {
    Object.entries(args).forEach(([key, val]) => {
      if (hasOwnProperty(config, key)) {
        config[key] = val;
      }
    });
  }
  if (!config.clientId || !config.clientSecret) {
    goodBye('Error: Missing client id or client secret');
  }

  if (typeof config.port === 'string') {
    // Convert to number
    config.port = +config.port;
  }

  return config;
};
