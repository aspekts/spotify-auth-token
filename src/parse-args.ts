import path from 'path';
import { SpotifyCreds } from './types';
import { read } from './utils';

export const parseArgs = () => {
  const credentialsFile = process.argv[2];
  if (!credentialsFile) {
    console.error('Error: No credentials file specified');
    process.exit(1);
  }
  const credentialsPath = path.join(process.cwd(), credentialsFile);
  const credentials = read<SpotifyCreds>(credentialsPath);
  if (!credentials.client_id || !credentials.client_secret) {
    console.error('Error: Invalid credentials file');
    process.exit(1);
  }
  return { credentials };
};
