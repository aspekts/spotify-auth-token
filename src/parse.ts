import path from 'path';
import { CLIInputConfig } from './types';
import { read } from './utils';

export const parseArgs = (): CLIInputConfig => {
  const optsFile = process.argv[2];

  if (!optsFile) {
    console.error('Error: No credentials file specified');
    process.exit(1);
  }
  const optsFilePath = path.join(process.cwd(), optsFile);
  const config = read<CLIInputConfig>(optsFilePath);
  if (!config.client_id || !config.client_secret) {
    console.error('Error: Invalid config file');
    process.exit(1);
  }
  return config;
};
