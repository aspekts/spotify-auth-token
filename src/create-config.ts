import { defaultConfig } from './config';
import { UserConfig } from './index';
import { AppConfig } from './types';
import { goodBye } from './utils';

function isNumeric(value: string) {
  return /^\d+$/.test(value);
}

export const createConfig = (args: UserConfig | string[]): AppConfig => {
  let config: Record<string, string | number> = {};

  // CLI use
  if (Array.isArray(args)) {
    config = args.reduce((acc, curr, idx, orig) => {
      if (curr.startsWith('--')) {
        const key = curr.slice(2);
        acc[key] = orig[idx + 1];
      }
      return acc;
    }, config);

    // Programmatic use
  } else {
    config = args;
  }
  if (!config.clientId || !config.clientSecret) {
    goodBye('Error: Missing client id or client secret');
  }

  if (typeof config.port === 'string') {
    // Convert to number
    config.port = +config.port;
  }

  return { ...defaultConfig, ...config } as AppConfig;
};
