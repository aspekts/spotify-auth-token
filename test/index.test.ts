import fs from 'fs';
import auth from '../src/index';
import * as request from '../src/request';
import { AppConfig } from '../src/types';
import * as utils from '../src/utils';

jest.useFakeTimers('modern').setSystemTime(new Date(1996, 3, 21));

const mockId = '1234';

jest.spyOn(utils, 'id').mockReturnValue(mockId);

const writeFileSpy = jest
  .spyOn(fs, 'writeFileSync')
  .mockImplementation(() => null);

const consoleInfoSpy = jest
  .spyOn(global.console, 'info')
  .mockImplementation(() => null);

const localhostSpy = jest
  .spyOn(request, 'getLocalhostUrl')
  .mockImplementation(() => {
    return Promise.resolve(`?code=AQDKHwNyRapw&state=${mockId}`);
  });

const requestSpy = jest.spyOn(request, 'request').mockImplementation(() => {
  return Promise.resolve({
    access_token: 'BQC2fMYf9',
    token_type: 'Bearer',
    expires_in: 3600,
    refresh_token: 'OFuVRQAzVzq0Wy_Py_W4KiMr8H0',
    scope: 'user-library-read',
  });
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('App', () => {
  const testConfigs: AppConfig[] = [
    {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 1,
      outDir: 'out/token',
      outFileName: 'mytoken',
      scopes: 'scopes1,scropes2',
    },
    {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 59,
      outDir: '/out/token/',
      outFileName: 'spotify-token',
      scopes: 'scopes1',
    },
    {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 69,
      outDir: '/out/',
      outFileName: 'spotify-token.json',
      scopes: 'scropes2',
    },
  ];

  testConfigs.forEach((config, idx) => {
    it(`works with config ${idx}`, async () => {
      await auth(config);

      expect(consoleInfoSpy.mock.calls[1][0]).toMatchSnapshot(
        `config ${idx}, auth url`
      );
      expect(localhostSpy).toHaveBeenCalledWith(config.port);
      expect(requestSpy.mock.calls[0][0]).toMatchSnapshot(
        `config ${idx}, spotify request`
      );
      const str = writeFileSpy.mock.calls[0][0] as string;
      const filePath = str.replace(/\\|\//g, ',').split(',').filter(Boolean);
      expect(filePath).toMatchSnapshot(`config ${idx}, out dir`);
      expect(
        JSON.parse(writeFileSpy.mock.calls[0][1] as string)
      ).toMatchSnapshot(`config ${idx}, written data`);
    });
  });
});
