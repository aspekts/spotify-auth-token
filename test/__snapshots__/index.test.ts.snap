// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App works with config 0: config 0, auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A1&scope=scopes1%252Cscropes2
"
`;

exports[`App works with config 0: config 0, out dir 1`] = `
Object {
  "base": "mytoken.json",
  "dir": "/usr/dir/out/token",
  "ext": ".json",
  "name": "mytoken",
  "root": "/",
}
`;

exports[`App works with config 0: config 0, spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`App works with config 0: config 0, written data 1`] = `
Object {
  "access_token": "BQC2fMYf9",
  "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
  "expires_in": 3600,
  "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
  "scope": "user-library-read",
  "token_type": "Bearer",
}
`;

exports[`App works with config 1: config 1, auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A59&scope=scopes1
"
`;

exports[`App works with config 1: config 1, out dir 1`] = `
Object {
  "base": "token.json",
  "dir": "/usr/dir/out/token",
  "ext": ".json",
  "name": "token",
  "root": "/",
}
`;

exports[`App works with config 1: config 1, spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`App works with config 1: config 1, written data 1`] = `
Object {
  "access_token": "BQC2fMYf9",
  "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
  "expires_in": 3600,
  "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
  "scope": "user-library-read",
  "token_type": "Bearer",
}
`;

exports[`App works with config 2: config 2, auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A69&scope=scropes2
"
`;

exports[`App works with config 2: config 2, out dir 1`] = `
Object {
  "base": "spotify-token.json",
  "dir": "/usr/dir/out",
  "ext": ".json",
  "name": "spotify-token",
  "root": "/",
}
`;

exports[`App works with config 2: config 2, spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`App works with config 2: config 2, written data 1`] = `
Object {
  "access_token": "BQC2fMYf9",
  "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
  "expires_in": 3600,
  "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
  "scope": "user-library-read",
  "token_type": "Bearer",
}
`;
